### ISSUE TITLE
live_slam crashes on startup

### ISSUE THREAD

**Issue by efernandez** [tum-vision/lsd_slam#3]
I'm using the `LSD_room.bag` dataset and running:

<pre>
rosrun lsd_slam_core live_slam image:=image_raw camera_info:=camera_info
</pre>

on a 32bit Ubuntu 12.04 computer with ROS hydro.

I'm having the following error, which seems to be related with Eigen (probably one of those MAKE_ALIGNED macros is missed in some clasess):

<pre>
WAITING for ROS camera calibration!
Received ROS Camera Calibration: fx: 254.326950, fy: 375.934387, cx: 267.381897, cy: 231.599091 @ 640x480
RECEIVED ROS camera calibration!
Started mapping thread!
Started  constraint search thread!
Started optimization thread 
Doing Random initialization!
live_slam: /usr/include/eigen3/Eigen/src/Core/DenseStorage.h:69: Eigen::internal::plain_array<T, Size, MatrixOrArrayOptions, 16>::plain_array() [with T = double, int Size = 4, int MatrixOrArrayOptions = 0]: Assertion `(reinterpret_cast<size_t>(array) & 0xf) == 0 && "this assertion is explained here: " "http://eigen.tuxfamily.org/dox-devel/TopicUnalignedArrayAssert.html" " **** READ THIS WEB PAGE !!! ****"' failed.
Aborted (core dumped)
</pre>

**Comment by efernandez** [comment]
Hereby I attach a backtrace:

<pre>
(gdb) bt
#0  0xb7fdd424 in __kernel_vsyscall ()
#1  0xb72ef1df in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xb72f2825 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0xb72e8085 in ?? () from /lib/i386-linux-gnu/libc.so.6
#4  0xb72e8137 in __assert_fail () from /lib/i386-linux-gnu/libc.so.6
#5  0xb7aa75fa in Eigen::internal::plain_array<double, 4, 0, 16>::plain_array (this=0x82f2a98)
    at /usr/include/eigen3/Eigen/src/Core/DenseStorage.h:69
#6  0xb7aa62f5 in Eigen::DenseStorage<double, 4, 4, 1, 0>::DenseStorage (this=0x82f2a98)
    at /usr/include/eigen3/Eigen/src/Core/DenseStorage.h:102
#7  0xb7aa3b24 in Eigen::PlainObjectBase<Eigen::Matrix<double, 4, 1, 0, 4, 1> >::PlainObjectBase (this=0x82f2a98)
    at /usr/include/eigen3/Eigen/src/Core/PlainObjectBase.h:395
#8  0xb7b2505f in Eigen::Matrix<double, 4, 1, 0, 4, 1>::Matrix (this=0x82f2a98, x=@0xbfffd500: 0, 
    y=@0xbfffd4f8: 0, z=@0xbfffd4f0: 0, w=@0xbfffd508: 1) at /usr/include/eigen3/Eigen/src/Core/Matrix.h:272
#9  0xb7b23e52 in Eigen::Quaternion<double, 0>::Quaternion (this=0x82f2a98, w=1, x=0, y=0, z=0)
    at /usr/include/eigen3/Eigen/src/Geometry/Quaternion.h:258
#10 0xb7b25a58 in Sophus::RxSO3Group<double, 0>::RxSO3Group (this=0x82f2a98)
    at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/thirdparty/Sophus/sophus/rxso3.hpp:625
#11 0xb7b24c8a in Sophus::Sim3Group<double, 0>::Sim3Group (this=0x82f2a98)
    at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/thirdparty/Sophus/sophus/sim3.hpp:697
#12 0xb7a99f39 in lsd_slam::Frame::Frame (this=0x82f2a88, id=2, width=640, height=480, K=..., 
    timestamp=0.45552300000000001, 
    image=0xaef96980 "\242\241\243\243\242\243\243\242\244\244\243\243\244\244\244\245\246\246\247\246\250\246\247\247\247\247\250\245\246\250\250\247\251\253\250\251\252\251\252\252\252\253\254\254\254\254\255\255\253\252\256\256\256\256\256\254\256\254\254\257\257\260\255\260\260\261\261\256\257\256\260\257\262\261\261\261\262\261\260\263\263\263\263\262\262\263\262\262\263\262\261\262\263\265\264\261\262\263\264\263\263\261\261\263\265\264\263\261\261\262\265\262\262\261\263\264\264\262\261\263\262\260\261\260\260\260\261\257\257\257\257\260\260\262\257\257\260\255\255\254\253\254\254\252\251\251\247\247\250\247\246\246\246\243\243\241\241\241\237\237\233\235\234\231\231\230\227\226\225\225\224\222\221\222\220\217\214\216\216\213\213\212\213\211\212\207\210\207\210\207\210\207\206\206\207\210\207\210\207\206"...) at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/src/DataStructures/Frame.cpp:35
#13 0xb7add6b6 in std::_Sp_counted_ptr_inplace<lsd_slam::Frame, std::allocator<lsd_slam::Frame>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<unsigned int&, int&, int&, Eigen::Matrix<float, 3, 3>&, double&, unsigned char*&>
    (this=0x82f2a68, __a=...) at /usr/include/c++/4.6/bits/shared_ptr_base.h:410
#14 0xb7ada4a9 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<lsd_slam::Frame, std::allocator<lsd_slam::Frame>, unsigned int&, int&, int&, Eigen::Matrix<float, 3, 3>&, double&, unsigned char*&> (
    this=0xbfffdb48, __a=...) at /usr/include/c++/4.6/bits/shared_ptr_base.h:518
#15 0xb7ad6688 in std::__shared_ptr<lsd_slam::Frame, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<lsd_slam::Frame>, unsigned int&, int&, int&, Eigen::Matrix<float, 3, 3>&, double&, unsigned char*&> (this=0xbfffdb44, 
    __tag=..., __a=...) at /usr/include/c++/4.6/bits/shared_ptr_base.h:987
#16 0xb7ad1542 in std::shared_ptr<lsd_slam::Frame>::shared_ptr<std::allocator<lsd_slam::Frame>, unsigned int&, int&, int&, Eigen::Matrix<float, 3, 3>&, double&, unsigned char*&> (this=0xbfffdb44, __tag=..., __a=...)
    at /usr/include/c++/4.6/bits/shared_ptr.h:317
#17 0xb7acbc0e in std::allocate_shared<lsd_slam::Frame, std::allocator<lsd_slam::Frame>, unsigned int&, int&, int&, Eigen::Matrix<float, 3, 3, 0, 3, 3>&, double&, unsigned char*&> (__a=...)
    at /usr/include/c++/4.6/bits/shared_ptr.h:535
#18 0xb7ac6ed0 in std::make_shared<lsd_slam::Frame, unsigned int&, int&, int&, Eigen::Matrix<float, 3, 3, 0, 3, 3>&, double&, unsigned char*&> () at /usr/include/c++/4.6/bits/shared_ptr.h:551
#19 0xb7abe33c in lsd_slam::SlamSystem::trackFrame (this=0x8150c10, 
    image=0xaef96980 "\242\241\243\243\242\243\243\242\244\244\243\243\244\244\244\245\246\246\247\246\250\246\247\247\247\247\250\245\246\250\250\247\251\253\250\251\252\251\252\252\252\253\254\254\254\254\255\255\253\252\256\256\256\256\256\254\256\254\254\257\257\260\255\260\260\261\261\256\257\256\260\257\262\261\261\261\262\261\260\263\263\263\263\262\262\263\262\262\263\262\261\262\263\265\264\261\262\263\264\263\263\261\261\263\265\264\263\261\261\262\265\262\262\261\263\264\264\262\261\263\262\260\261\260\260\260\261\257\257\257\257\260\260\262\257\257\260\255\255\254\253\254\254\252\251\251\247\247\250\247\246\246\246\243\243\241\241\241\237\237\233\235\234\231\231\230\227\226\225\225\224\222\221\222\220\217\214\216\216\213\213\212\213\211\212\207\210\207\210\207\210\207\206\206\207\210\207\210\207\206"..., frameID=2, blockUntilMapped=false, timestamp=0.45552300000000001)
    at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/src/SlamSystem.cpp:890
#20 0xb7b0bb80 in lsd_slam::LiveSLAMWrapper::newImageCallback (this=0xbfffe440, img=..., imgTime=...)
    at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/src/LiveSLAMWrapper.cpp:137
---Type <return> to continue, or q <return> to quit---
#21 0xb7b0b937 in lsd_slam::LiveSLAMWrapper::Loop (this=0xbfffe440)
    at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/src/LiveSLAMWrapper.cpp:107
#22 0x080a5592 in main (argc=1, argv=0xbfffe584)
    at /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/src/main_live_odometry.cpp:61
</pre>

**Comment by efernandez** [comment]
Apparently it's a limitation of `boost::make_shared`: http://lists.boost.org/boost-users/2010/03/57713.php

Therefore, I've replaced them with `new`, using `reset` if needed, in #4.

But now I have a SIGSEGV:

<pre>
#0  0xb7d660ce in lsd_slam::FramePoseStruct::FramePoseStruct(lsd_slam::Frame*) ()
   from /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/lib/liblsdslam.so
#1  0xb7d60169 in lsd_slam::Frame::initialize(int, int, int, Eigen::Matrix<float, 3, 3, 0, 3, 3> const&, double) ()
   from /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/lib/liblsdslam.so
#2  0xb7d611cd in lsd_slam::Frame::Frame(int, int, int, Eigen::Matrix<float, 3, 3, 0, 3, 3> const&, double, unsigned char const*) () from /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/lib/liblsdslam.so
#3  0xb7d6fdf3 in lsd_slam::SlamSystem::trackFrame(unsigned char*, unsigned int, bool, double) ()
   from /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/lib/liblsdslam.so
#4  0xb7d86d35 in lsd_slam::LiveSLAMWrapper::newImageCallback(cv::Mat const&, lsd_slam::Timestamp) ()
   from /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/lib/liblsdslam.so
#5  0xb7d883e0 in lsd_slam::LiveSLAMWrapper::Loop() ()
   from /home/enriquefernandez/svn/stacks/lsd_slam/lsd_slam_core/lib/liblsdslam.so
#6  0x0805ede7 in main ()
</pre>

**Comment by JakobEngel** [comment]
thanks for the PR's - I never got around to fixing those warnings.
what's that SIGSEGV / is it still a problem?

**Comment by efernandez** [comment]
@artivis and me are working on it.

So far, we've disable SSE and fix a lot of issues with EIGEN unligned memory.
@artivis will send another PR that fix all of them, but with SSE disable we have like 1Hz framerate with the `LSD_room.bag`; is that normal? I mean, with SSE enabled which framerate should I expect to have?

For me there's no segfault without SSE, but @artivis is having problems with GTK (I think); he might try to follow @mhkabir advise in #1 (@mhkabir could you send a PR at least only for that part, of the needed changes)?
Thanks!

**Comment by versatran01** [comment]
I experienced a core dumped problem as well. It works for a few frames at 1hz and then segmentation fault.
 I tried building with different optimization levels and enabled/disabled sse. Doesn't seem to solve the problem.

and the live slam keeps saying:
virtual bool g2o::SparseOptimizer::initializeOptimization(g2o::HyperGraph::VertexSet&, int): Attempt to initialize an empty graph
virtual int g2o::SparseOptimizer::optimize(int, bool): 0 vertices to optimize, maybe forgot to call initializeOptimization()

I build g2o from source instead of installing from apt-get.
An ideas on this?
